source/server/connection_handler_impl.cc

```diff
@@ -77,15 +77,29 @@ ConnectionHandlerImpl::SslActiveListener::SslActiveListener(
                                                            parent.stats_store_, listener_options),
                      factory, socket.localAddress()->asString()) {}
 
-Network::Listener* ConnectionHandlerImpl::findListenerByPort(uint32_t port) {
-  // findListenerByPort is O(#listeners) operation, may need to add a map<port, listener>
-  // to improve performance to O(log #listeners); since the number of listeners is small
-  // linear performance might be adequate
-  auto l = std::find_if(
-      listeners_.begin(), listeners_.end(),
-      [port](const std::pair<Network::Address::InstancePtr, ActiveListenerPtr>& p) {
-        return p.first->type() == Network::Address::Type::Ip && p.first->ip()->port() == port;
-      });
+Network::Listener*
+ConnectionHandlerImpl::findListenerByAddress(Network::Address::InstancePtr address) {
+  // This is a linear operation, may need to add a map<address, listener> to improve performance.
+  // However, linear performance might be adequate since the number of listeners is small.
+  auto l =
+      std::find_if(listeners_.begin(), listeners_.end(),
+                   [address](const std::pair<Network::Address::InstancePtr, ActiveListenerPtr>& p) {
+                     return p.first->type() == Network::Address::Type::Ip &&
+                            p.first->asString() == address->asString();
+                   });
+
+  // If there is exact address match, return the corresponding listener.
+  if (l != listeners_.end()) {
+    return l->second->listener_.get();
+  }
+
+  // Otherwise, we need to look for the wild card match, i.e., 0.0.0.0:[address_port].
+  l = std::find_if(listeners_.begin(), listeners_.end(),
+                   [address](const std::pair<Network::Address::InstancePtr, ActiveListenerPtr>& p) {
+                     return p.first->type() == Network::Address::Type::Ip &&
+                            p.first->ip()->port() == address->ip()->port() &&
+                            p.first->ip()->addressAsString() == "0.0.0.0";
```

Comment:
Maybe we should have an attribute on Address that lets us know if it's 0.0.0.0 without having to do string compares?